milabench/metrics/sqlalchemy.py

```diff
@@ -0,0 +1,350 @@
+            namespace=None,
+            created_time=datetime.utcnow(),
+            meta=entry.data,
+            status="running",
+        )
+        self.session.add(self.run)
+        self.session.commit()
+        self.session.refresh(self.run)
+
+    def on_new_pack(self, entry):
+        state = self.pack_state(entry)
+        state.pack = Pack(
+            exec_id=self.run._id,
+            created_time=datetime.utcnow(),
+            name=entry.pack.config["name"],
+            tag=entry.tag,
+            config=entry.pack.config,
+        )
+        self.session.add(state.pack)
+        self.session.commit()
+        self.session.refresh(state.pack)
+
+    def on_meta(self, entry: BenchLogEntry):
+        if self.run is None:
+            self.on_new_run(entry)
+
+        if entry.tag not in self.states:
+            self.on_new_pack(entry)
+
+        state = self.pack_state(entry)
+        assert state.step == META
+        state.step += 1
+
+    def on_start(self, entry):
+        state = self.pack_state(entry)
+        state.pack.command = entry.data["command"]
+        state.start = entry.data["time"]
+
+        assert state.step == START
+        state.step += 1
+
+    def on_phase(self, entry):
+        pass
+
+    def on_error(self, entry):
+        state = self.pack_state(entry)
+        state.error += 1
+
+    def on_line(self, entry):
+        pass
+
+    def _push_metric(self, run_id, pack_id, name, value, gpu_id=None):
+        self.pending_metrics.append(
+            Metric(
+                exec_id=run_id,
+                pack_id=pack_id,
+                name=name,
+                value=value,
+                gpu_id=gpu_id,
+            )
+        )
+
+    def _change_gpudata(self, run_id, pack_id, k, v, jobid):
+        for gpu_id, values in v.items():
+            for metric, value in values.items():
+                if metric == "memory":
+                    use, mx = value
+                    value = use / mx
+
+                assert jobid == gpu_id, f"{type(jobid)} == {type(gpu_id)}"
+                self._push_metric(run_id, pack_id, f"gpu.{metric}", value, gpu_id)
+
+    def on_data(self, entry):
+        state = self.pack_state(entry)
+        assert state.step == DATA
+
+        run_id = self.run._id
+        pack_id = state.pack._id
+        jobid = str(state.pack.config["job-number"])
+
+        for k, v in entry.data.items():
+            if k in ("task", "units", "progress"):
+                continue
+
+            # GPU data would have been too hard to query
+            # so the gpu_id is moved to its own column
+            # and each metric is pushed as a separate document
+            if k == "gpudata":
+                self._change_gpudata(run_id, pack_id, k, v, jobid)
+                return
+
+            # We request an ordered write so the document
+            # will be ordered by their _id (i.e insert time)
+            self._push_metric(run_id, pack_id, k, v, gpu_id=jobid)
```

Comment:
Might want to check that `v` is int or float.

Reply:
done, checks that the value is numeric inside the _push_metric